src/porepy/numerics/fv/tpsa.py

```diff
@@ -0,0 +1,1379 @@
+        n_fi = sd.face_normals[:, numbering.fi]
+        # Switch signs where relevant
+        n_fi *= numbering.sgn
+
+        # Get a vector from cell center to face center and project to the direction of
+        # the face normal. Divide by the face area to get a unit vector in the normal
+        # direction.
+        fc_cc = (
+            n_fi
+            * (sd.face_centers[::, numbering.fi] - sd.cell_centers[::, numbering.ci])
+            / sd.face_areas[numbering.fi]
+        )
+        # Get the length of the projected vector; take the absolute value to avoid
+        # negative distances.
+        dist_fc_cc = np.abs(np.sum(fc_cc, axis=0))
+
+        # Construct mu_i / delta_k^i, and its nd version
+        mu_by_dist_fc_cc = mu / dist_fc_cc
+        mu_by_dist_fc_cc_nd = np.repeat(mu_by_dist_fc_cc, nd)
+
+        # Extract the diagonal of the Robin weight. Non-diagonal elements are ignored
+        # here and specifically ruled out (with error messages) elsewhere in the code.
+        rob_weight = np.vstack(
+            (bnd_disp.robin_weight[0, 0], bnd_disp.robin_weight[1, 1])
+        )
+        if nd == 3:
+            rob_weight = np.vstack((rob_weight, bnd_disp.robin_weight[2, 2]))
+
+        # Nd version of the faces with Robin boundary conditions.
+        rob_boundary_faces_expanded = pp.fvutils.expand_indices_nd(
+            np.arange(nf), nd
+        ).reshape((nd, nf), order="F")[bnd_disp.is_rob]
+        rob_weights_boundary_faces = rob_weight[bnd_disp.is_rob]
+
+        # This is the face-wise sum of the expressions mu/delta, also accounting for
+        # boundary conditions. EK note to self: The factor 2 is motivated by Robin
+        # boundary conditions: With the Robin weight represented, PorePy style, as
+        # \alpha (rather than 2*\mu*\delta^-1 as is used in the paper), the averagaging
+        # operator for a face with a Robin condition becomes
+        # (2*\mu*\delta^-1)/(2*\mu*\delta^-1 + \alpha) for the cell contribution; the
+        # contribution from the boundary condition has \alpha in the nominator. The
+        # factor 2 introduced here, and inherited in the diagonal scaling matrix
+        # constructed next, must be compensated by a factor 2 in the nominator of the
+        # averaging map, as is done elsewhere in the code.
+        #
+        # For reference, the reciprocal of this field is also the expression \delta_k^mu
+        # (used in the paper to describe the discretization scheme).
+        mu_by_dist_fc_cc_bound = np.bincount(
+            np.hstack((numbering.fi_expanded, rob_boundary_faces_expanded)),
+            weights=np.hstack((2 * mu_by_dist_fc_cc_nd, rob_weights_boundary_faces)),
+        )
+
+        # Create a diagonal matrix that can be used to scale face-to-cell maps to have
+        # unit row sum (thus they become true averaging maps), both in the interior and
+        # on faces with Robin boundary conditions.
+        inv_mu_by_dist_array = sps.dia_array(
```

Comment:
I don't think that would work, since we need to left-multiply various matrices with this variable. I am not in complete control of the nuances in scipy-numpy compatability, and associated broadcasting, so my preference is to make it safe.